relay-general/src/store/transactions/processor.rs

```diff
@@ -1,21 +1,89 @@
+use chrono::Utc;
+
 use relay_common::SpanStatus;
 
 use crate::processor::{ProcessValue, ProcessingState, Processor};
 use crate::protocol::{
-    Context, ContextInner, Event, EventType, Span, Timestamp, TransactionSource,
+    Context, ContextInner, Event, EventType, Span, Timestamp, TransactionInfo, TransactionSource,
 };
 use crate::store::regexes::TRANSACTION_NAME_NORMALIZER_REGEX;
 use crate::types::{Annotated, Meta, ProcessingAction, ProcessingResult, Remark, RemarkType};
 
+use super::TransactionNameRule;
+
 /// Rejects transactions based on required fields.
 #[derive(Default)]
-pub struct TransactionsProcessor {
+pub struct TransactionsProcessor<'r> {
     normalize_names: bool,
+    tx_name_rules: &'r [TransactionNameRule],
 }
 
-impl TransactionsProcessor {
-    pub fn new(normalize_names: bool) -> Self {
-        Self { normalize_names }
+impl<'r> TransactionsProcessor<'r> {
+    pub fn new(normalize_names: bool, tx_name_rules: &'r [TransactionNameRule]) -> Self {
+        Self {
+            normalize_names,
+            tx_name_rules,
+        }
+    }
+
+    /// Applies the rule if any found to the transaction name.
+    ///
+    /// It find the first rule matching the criteria:
+    /// - source matchining the one provided in the rule sorce, default `url`
+    /// - rule hasn't epired yet
+    /// - glob pattern matches the transaction name
+    ///
+    /// Note: we add `/` at the end of the transaction name if there isn't one, to make sure that
+    /// patterns like `/<something>/*/**` where we have `**` at the end are a match.
+    pub fn apply_transaction_rename_rule(
+        &self,
+        transaction: &mut Annotated<String>,
+        info: &mut std::option::Option<TransactionInfo>,
+        source: &TransactionSource,
+    ) -> ProcessingResult {
+        let now = Utc::now();
+        transaction.apply(|transaction, meta| {
+            let slash_is_present = transaction
+                .chars()
+                .last()
+                .map(|c| c == '/')
+                .unwrap_or_default();
+
+            // Add new `/` at the end of the transaction if there isn't one.
+            if !slash_is_present {
+                transaction.push('/');
+            }
+
+            let rule = self.tx_name_rules.iter().find(|rule| {
+                &rule.scope.source == source
+                    && rule.expiry > now
+                    // Adding `/` at the end of the name, ensures that rules like /<something>/*/**
+                    // will always match the string.
+                    && rule.pattern.is_match(transaction)
+            });
+            let result = rule.map(|r| (r.pattern.pattern(), r.apply(transaction)));
+
+            if let Some((rule, Some(mut result))) = result {
+                if !slash_is_present {
+                    let _ = transaction.pop();
+                    let _ = result.pop();
```

Comment:
```suggestion
                    transaction.pop();
                    result.pop();
```
Why not? Same on line 83.